blacktie/jatmibroker-xatmi/src/main/java/org/jboss/narayana/blacktie/jatmibroker/core/transport/hybrid/stomp/StompManagement.java

```diff
@@ -67,28 +67,30 @@ public static Socket connect(String host, int port, String username, String pass
     }
 
     public static void send(Message message, OutputStream outputStream) throws IOException {
-        log.trace("Writing on: " + outputStream);
-        StringBuffer toSend = new StringBuffer(message.getCommand().toString());
-        toSend.append("\n");
-
-        Iterator<String> keys = message.getHeaders().keySet().iterator();
-        while (keys.hasNext()) {
-            String key = keys.next();
-            String value = message.getHeaders().get(key);
-            toSend.append(key);
-            toSend.append(":");
-            toSend.append(value);
+        synchronized (outputStream) {
```

Comment:
I also think it is best to assemble the data into the buffer and then write it to the socket in one go. So using getChannel and assembling the data before writing we would have:

        ByteBuffer toSend = ByteBuffer.wrap(message.getCommand().getBytes());

        if (message.getBody() != null)
            toSend.put(message.getBody());
     
        toSend.put("\000\n\n".getBytes()); 
       
        socket.getChannel().write(toSend);

Implementing in this way avoids the need for synchronization.